src/wp-includes/user.php

```diff
@@ -3012,11 +3012,55 @@ function retrieve_password( $user_login = null ) {
 	 */
 	$message = apply_filters( 'retrieve_password_message', $message, $key, $user_login, $user_data );
 
+	// Wrap the single notification email arguments in an array to pass them to the retrieve_password_notification_email filter.
+	$defaults = array(
+		'to'      => $user_email,
+		'subject' => $title,
+		'message' => $message,
+		'headers' => '',
+	);
+
+	$data = compact( 'key', 'user_login', 'user_data' );
+
+	/**
+	 * Filter the contents of the reset password notification email sent to the user.
+	 *
+	 * @since 6.0.0
+	 *
+	 * @param array   $defaults {
+	 *     The default notification email arguments. Used to build wp_mail().
+	 *
+	 *     @type string $to      The intended recipient - user email address.
+	 *     @type string $subject The subject of the email.
+	 *     @type string $message The body of the email.
+	 *     @type string $headers The headers of the email.
+	 * }
+	 * @param array $data {
+	 *     Additional information for extenders.
+	 *
+	 *     @type string  $key        The activation key.
+	 *     @type string  $user_login The username for the user.
+	 *     @type WP_User $user_data  WP_User object.
+	 * }
+	 */
+	$notification_email = apply_filters( 'retrieve_password_notification_email', $defaults, $data );
+
 	if ( $switched_locale ) {
 		restore_previous_locale();
 	}
 
-	if ( $message && ! wp_mail( $user_email, wp_specialchars_decode( $title ), $message ) ) {
+	if ( is_array( $notification_email ) ) {
+		// Force key order and merge defaults in case any value is missing in the filtered array.
+		$notification_email = array_merge( $defaults, $notification_email );
+	} else {
+		$notification_email = $defaults;
+	}
+
+	list( $to, $subject, $message, $headers ) = array_values( $notification_email );
+
+	$subject = wp_specialchars_decode( $subject );
+
+	if ( ! wp_mail( $to, $subject, $message, $headers ) ) {
```

Comment:
This appears to have changed the logic around whether the email gets sent. Previously, an empty return value from `retrieve_password_message` prevents `wp_mail()` from being called:

```php
if ( $message && ! wp_mail( ... ) ) {
```

([This behaviour is documented here](https://github.com/johnbillion/wp_mail/#user-or-administrator-requests-a-password-reset)).

With this change in place `wp_mail()` is always called, potentially with an empty message body.

Reply:
Aha, it's there to disable the retrieve password email. It looked like a silently unsent email was unintended. Thanks @johnbillion!

An empty `message` can exist for a number of reasons and this falsey check unfortunately means it will continue to fail silently when 'retrieve_password_message' returns an empty `message`. PHPMailer has an error for an empty message and we don't have `$allowEmpty` set to `true`.

For performance and simplicity, if we're going to disable the email, we should do so as early as possible in the `retrieve_password()` function.

Given the above and considering backwards compatibility, I suggest we make the following changes, including an additional new filter:

```php
...

// At L2774, just after the `if ( ! $user_data ) {` block.

/**
 * Filter whether to disable the retrieve password email.
 *
 * @since 6.0.0
 *
 * @param bool $disable True to disable. Default: false
*/
if ( apply_filters( 'disable_retrieve_password_email', false ) ) {
    return true;
}

...

$message = apply_filters( 'retrieve_password_message'...

// Short-circuit on falsey $message value for backwards compatibility.
if ( ! $message ) {
    return true;
}

...
```

This means:

- we maintain BC.
- the docs for `retrieve_password_message` do not need to change.
- we introduce explicit disabling of the email with `(bool) true` in the new filter.
- PHPMailer will error out if the `$notification_email` message is falsey. We should not be failing silently here.
- if performance is a concern due to calling `wp_mail()` with an empty message, we _should_ have a guard at the start of `wp_mail()` to error out in _all_ instances of an empty message.

In terms of scope:
- The ticket is about one new filter.
- These changes are about how we can implement this new filter without repeating the falsey check. Instead, by making the disable explicit, this prevents unintentionally empty messages from failing silently.

However, I'm happy to open a separate ticket if that's preferred.

Thoughts @johnbillion, @MarloKessler?

Reply:
Sounds good to me!

Reply:
@johnbillion I implemented the solution of @costdev. Is this okay for you?